### PR TITLE
[C2y] Claim partial conformance to WG14 N3244

### PR DESCRIPTION
**Pull request by AaronBallman**
This paper had several changes within it, and Clang implements some of the changes, but not others. This updates the status accordingly.

### CODE REVIEW

**Comment by AaronBallman** [comment]
The paper in question can be found at: https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3244.pdf

File: clang/test/C/C2y/n3244.c [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+_Static_assert(_Alignof(AlignmentOnOriginalDecl) == 8, "");
+
+long long CompatibleAlignment;
+_Static_assert(_Alignof(CompatibleAlignment) == _Alignof(long long), "");
+_Alignas(_Alignof(long long)) long long CompatibleAlignment; // Okay, alignment is the same as the implied alignment
+
+_Alignas(_Alignof(long long)) long long CompatibleAlignment2;
+// FIXME: this should be accepted because the redeclaration has no alignment
+// specifier.
+long long CompatibleAlignment2; // expected-error {{'_Alignas' must be specified on definition if it is specified on any declaration}}
```
Comment (efriedma-quic) [inline clang/test/C/C2y/n3244.c:65]: All these testcases involve tentative definitions... maybe worth adding a few that have "extern" declarations or explicit definitions?

Comment (AaronBallman) [reply clang/test/C/C2y/n3244.c:65]: Sure! I could also add some explicitly incompatible alignments as well.

**Review by efriedma-quic — APPROVED** [review]
LGTM

**Comment by mgorny** [comment]
The added test fails on 32-bit x86:

```
FAIL: Clang :: C/C2y/n3244.c (1461 of 20066)
******************** TEST 'Clang :: C/C2y/n3244.c' FAILED ********************
Exit Code: 1

Command Output (stderr):
--
RUN: at line 1: /var/tmp/portage/sys-devel/clang-19.0.0_pre20240713/work/x/y/clang-abi_x86_32.x86/bin/clang -cc1 -internal-isystem /var/tmp/portage/sys-devel/clang-19.0.0_pre20240713/work/x/y/clang-abi_x86_32.x86/bin/../../../../lib/clang/19/include -nostdsysteminc -std=c2y /var/tmp/portage/sys-devel/clang-19.0.0_pre20240713/work/clang/test/C/C2y/n3244.c -verify -Wno-gnu-alignof-expression
+ /var/tmp/portage/sys-devel/clang-19.0.0_pre20240713/work/x/y/clang-abi_x86_32.x86/bin/clang -cc1 -internal-isystem /var/tmp/portage/sys-devel/clang-19.0.0_pre20240713/work/x/y/clang-abi_x86_32.x86/bin/../../../../lib/clang/19/include -nostdsysteminc -std=c2y /var/tmp/portage/sys-devel/clang-19.0.0_pre20240713/work/clang/test/C/C2y/n3244.c -verify -Wno-gnu-alignof-expression
error: 'expected-error' diagnostics seen but not expected: 
  File /var/tmp/portage/sys-devel/clang-19.0.0_pre20240713/work/clang/test/C/C2y/n3244.c Line 59: static assertion failed due to requirement '_Alignof (CompatibleAlignment) == _Alignof(long long)': 
error: 'expected-note' diagnostics seen but not expected: 
  File /var/tmp/portage/sys-devel/clang-19.0.0_pre20240713/work/clang/test/C/C2y/n3244.c Line 59: expression evaluates to '8 == 4'
2 errors generated.

--

********************
```

**Comment by AaronBallman** [comment]
Thank you for letting me know, I've pushed a fix for that.

Curiously, it seems `_Alignof` on an expression always returns the *preferred* alignment and not the actual alignment. This is consistent with GCC, even if it's a bit surprising: https://godbolt.org/z/nvKx98ff9

**Comment by mgorny** [comment]
Thanks. I can confirm that the tests pass for me with the patch applied.

**Pull request merged**